packages/falter-berlin-hna-list/src/libhna/hna.c

```diff
@@ -0,0 +1,169 @@
+                                }
+                                tok_end += 1;
+                        }
+                        memcpy(name, tok_beg, tok_end - tok_beg);
+                        name[(tok_end - tok_beg)] = '\0';
+                        //printf("found: '%.*s'\n", (int) (tok_end - tok_beg), tok_beg);
+                        return 1;
+                }
+        }
+        return 0;
+}
+
+char *tcp_read(const char *addr_str, int port, const char *request) {
+        struct sockaddr_storage server_addr;
+
+        size_t buffer_len = 4096 * 4;
+        char *buffer = malloc(buffer_len);
+        int sock = socket(AF_INET, SOCK_STREAM, 0);
+
+        if (sock < 0) {
+                printf("socket(): %s\n", strerror(errno));
+                return NULL;
+        }
+
+        struct sockaddr_in *addr4 = (struct sockaddr_in *) &server_addr;
+        struct sockaddr_in6 *addr6 = (struct sockaddr_in6 *) &server_addr;
+
+        if (inet_pton(AF_INET, addr_str, &addr4->sin_addr) == 1) {
+                addr4->sin_family = AF_INET;
+                addr4->sin_port = htons(port);
+        } else if (inet_pton(AF_INET6, addr_str, &addr6->sin6_addr) == 1) {
+                addr6->sin6_family = AF_INET6;
+                addr6->sin6_port = htons(port);
+        } else {
+                printf("invalid address: %s\n", addr_str);
+                return NULL;
+        }
+
+        if (connect(sock, (struct sockaddr *) &server_addr, sizeof(server_addr)) < 0) {
+                printf("connect(): %s\n", strerror(errno));
+                return NULL;
+        }
+
+        send(sock, request, strlen(request), 0);
+        int valread = read(sock, buffer, buffer_len - 1);
+        if (valread > 0) {
+                return buffer;
+        } else {
+                return NULL;
+        }
+}
+
+
+hna_data *serialize_hna_string(char *hna, char *gateway) {
+        // input-data: "10.230.198.192/28", "10.31.43.188"
+        hna_data* dataset = calloc(1, sizeof(hna_data));
+
+        char* ipaddr = strtok(hna, "/");
+        char* netmask = strtok(NULL, "/");
+
+        //ToDo: struct sockaddr_in* addr_struct = (struct sockaddr_in)dataset->hna.base_addr;
+        inet_pton(AF_INET, ipaddr, &dataset->hna.base_addr.sin_addr);
+        dataset->hna.netmask = netmask;
+
+        inet_pton(AF_INET, gateway, &dataset->via_gateway.sin_addr);
+
+        return dataset;
+}
+
+void read_hna_into_tree(struct avl_table *tree, char *raw_data) {
+        char *str1, *str2, *token, *subtoken;
+        char *saveptr1, *saveptr2;
+        char *line_delim = "\n";
+        char *field_delim = "\t";
+        int j;
+        for (str1 = raw_data;;str1 = NULL) {
+                token = strtok_r(str1, line_delim, &saveptr1);
+                if (token == NULL) {
+                        break;
+                }
+                if (token[0] == 'T' || token[0] == 'D') {
+                        continue;
+                }
+
+                char* args[2];
```

Comment:
char* args[2] = {NULL, NULL};

^ in general a good idea

Reply:
ammended in recent version.